hail/Makefile

```diff
@@ -66,26 +67,56 @@ wheel: build-info shadowJar init-scripts
 	cp build/deploy/src/hail/hail_version build/deploy/src/hailctl/hail_version
 	cd build/deploy; $(HAIL_PYTHON3) setup.py sdist bdist_wheel
 
-install-wheel: wheel
-	pip uninstall -y hail
-	pip install build/deploy/dist/*.whl
-
 # if the DEPLOY_PROD flag is not set, then deploy init scripts into a dev-username location
 ifndef DEPLOY_PROD
-DEV_SALT := $(shell whoami)-dev
+DEV_CLARIFIER := $(shell whoami)-dev
+CLOUD_SUB_FOLDER := $(HAIL_PIP_VERSION)-$(SHORT_REVISION)
+else
+CLOUD_SUB_FOLDER := $(HAIL_PIP_VERSION)
 endif
 
-CLOUD_LOC = gs://hail-common/hailctl/dataproc/$(DEV_SALT)/$(HAIL_PIP_VERSION)/$(SHORT_REVISION)/$(FILENAME)
-RESOURCES_TO_COPY = $(shell ls python/hailctl/dataproc/resources)
+HAILCTL_BUCKET_BASE ?= gs://hail-common/hailctl/dataproc
+
+cloud_base := $(HAILCTL_BUCKET_BASE)/$(DEV_CLARIFIER)/$(CLOUD_SUB_FOLDER)
+jar_cloud_path := $(cloud_base)/hail.jar
+wheel_cloud_path := $(cloud_base)/hail-$(HAIL_PIP_VERSION)-py3-none-any.whl
+wheel_path := build/deploy/dist/hail-$(HAIL_PIP_VERSION)-py3-none-any.whl
+.PHONY: python/hailctl/deploy.yaml
 python/hailctl/deploy.yaml:
 	rm -f python/hailctl/deploy.yaml
 	echo "dataproc:" >> python/hailctl/deploy.yaml
-	$(foreach FILENAME, $(RESOURCES_TO_COPY), \
-	    gsutil cp python/hailctl/dataproc/resources/$(FILENAME) $(CLOUD_LOC); \
-	    gsutil acl set public-read $(CLOUD_LOC); \
-	    echo "  $(FILENAME): $(CLOUD_LOC)" >> python/hailctl/deploy.yaml;)
+	for FILE in `ls python/hailctl/dataproc/resources`; do \
+	  CLOUD_LOC=$(HAILCTL_BUCKET_BASE)/$(DEV_CLARIFIER)/$(CLOUD_SUB_FOLDER)/$$FILE && \
+	  gsutil -m cp python/hailctl/dataproc/resources/$$FILE $$CLOUD_LOC && \
+	  gsutil acl set public-read $$CLOUD_LOC && \
+	  echo "  $$FILE: $$CLOUD_LOC" >> python/hailctl/deploy.yaml || break; done || exit
+	echo "  jar: $(jar_cloud_path)" >> python/hailctl/deploy.yaml
+	echo "  wheel: $(wheel_cloud_path)" >> python/hailctl/deploy.yaml
+
+.PHONY: upload-deferred-artifacts
+upload-deferred-artifacts: wheel
+	gsutil -m cp build/libs/hail-all-spark.jar $(jar_cloud_path)
+	gsutil acl set public-read $(jar_cloud_path)
+	gsutil -m cp $(wheel_path) $(wheel_cloud_path)
+	gsutil acl set public-read $(wheel_cloud_path)
+
+.PHONY: install-wheel
+install-wheel: wheel
+	pip uninstall -y hail
+	pip install $(wheel_path)
 
-.PHONY: pypi-deploy
+.PHONY: install-hailctl-wheel
```

Comment:
I really like this style of PHONY next to the rule in question, but it, oddly, seems non-standard. ❤️

Reply:
I copied this style from some stuff I saw on SO and also in the autogenerated sphinx makefile. It does seem much clearer -- the clarifier is next to the code in question.